pkg/controllers/provisioning/scheduling/node.go

```diff
@@ -27,65 +27,133 @@ import (
 // Node is a set of constraints, compatible pods, and possible instance types that could fulfill these constraints. This
 // will be turned into one or more actual node instances within the cluster after bin packing.
 type Node struct {
-	Constraints         *v1alpha5.Constraints
+	v1alpha5.Constraints
 	InstanceTypeOptions []cloudprovider.InstanceType
 	Pods                []*v1.Pod
-	unschedulable       bool // if true, the node will accept no more pods
+
+	podResources    v1.ResourceList
+	daemonResources v1.ResourceList
+	daemons         []daemonInfo
+}
+
+type daemonInfo struct {
+	pod          *v1.Pod
+	requirements v1alpha5.Requirements
+	resources    v1.ResourceList
 }
 
-func NewNode(constraints *v1alpha5.Constraints, instanceTypeOptions []cloudprovider.InstanceType, pods ...*v1.Pod) *Node {
-	n := &Node{Constraints: constraints.DeepCopy()}
+func NewNode(constraints *v1alpha5.Constraints, daemons []*v1.Pod, instanceTypeOptions []cloudprovider.InstanceType, pods ...*v1.Pod) (*Node, error) {
+	n := &Node{
+		Constraints: *constraints.DeepCopy(),
+	}
+
+	for _, daemon := range daemons {
+		// skip any daemons that our provisioner configured taints would cause to not schedule
+		if err := n.Taints.Tolerates(daemon); err != nil {
+			continue
+		}
+		di := daemonInfo{
+			pod:          daemon,
+			requirements: v1alpha5.NewPodRequirements(daemon),
+			resources:    resources.RequestsForPods(daemon),
+		}
+		n.daemons = append(n.daemons, di)
+	}
+
+	// calculate daemon resource consumption so we can filter out instance types based on daemons + instance type overhead
+	n.recalculateDaemonResources()
 
 	for _, it := range instanceTypeOptions {
+		// If a zero-resource pod can't fit, don't consider this instance type.  This occurs if the node overhead +
+		// daemon set requests is already larger than the instance type can support
+		if !n.newPodCanFit(nil, it) {
+			continue
+		}
+
 		// pre-filter our list of all possible instance types by what the provisioner allows
 		if !cloudprovider.Compatible(it, constraints.Requirements) {
 			continue
 		}
 		n.InstanceTypeOptions = append(n.InstanceTypeOptions, it)
 	}
-	// technically we might create some invalid construct here if the pod can't be supported by any instance type
-	// but this is ok as we just won't have any valid instance types, this pod won't be schedulable and nothing else
-	// will be compatible with this node
+
 	for _, p := range pods {
 		n.Add(p)
 	}
-	return n
+	if len(n.InstanceTypeOptions) == 0 {
+		return nil, errors.New("no instance type satisfied requirements")
```

Comment:
Thoughts on including resources and requirements in this error message?

Reply:
Would probably be helpful, examples below:

```
 logger.go:130: 2022-03-26T21:53:30.822-0500	ERROR	provisioning	scheduling/scheduler.go:98	Scheduling pod, no instance type satisfied resources {memory: 2Ti} and requirements kubernetes.io/os In [darwin ios linux windows], karpenter.sh/capacity-type In [on-demand spot], node.kubernetes.io/instance-type In [small-instance-type], karpenter.sh/provisioner-name In [vultureglaze], topology.kubernetes.io/zone In [test-zone-1 test-zone-2 test-zone-3], kubernetes.io/arch In [amd64 arm64]	{"provisioner": "vultureglaze", "pod": "default/p0888-huggersand-025446"}

   logger.go:130: 2022-03-26T21:54:52.017-0500	ERROR	provisioning	scheduling/scheduler.go:98	Scheduling pod, no instance type satisfied resources {cpu: 14} and requirements karpenter.sh/provisioner-name In [fliermoor], topology.kubernetes.io/zone In [test-zone-1 test-zone-2 test-zone-3], kubernetes.io/os In [darwin ios linux windows], karpenter.sh/capacity-type In [on-demand spot], kubernetes.io/arch In [amd64], node.kubernetes.io/instance-type In [amd-gpu-instance-type arm-instance-type aws-neuron-instance-type default-instance-type nvidia-gpu-instance-type and 2 others]	{"provisioner": "fliermoor", "pod": "default/p0351-oxfast-040149"}
 
   ```